source/common/quic/quic_transport_socket_factory.h

```diff
@@ -28,19 +53,42 @@ class QuicTransportSocketFactoryBase : public Network::TransportSocketFactory {
   bool implementsSecureTransport() const override { return true; }
   bool usesProxyProtocolOptions() const override { return false; }
   bool supportsAlpn() const override { return true; }
+
+protected:
+  virtual void onSecretUpdated() = 0;
+  QuicTransportSocketFactoryStats stats_;
 };
 
 // TODO(danzh): when implement ProofSource, examine of it's necessary to
 // differentiate server and client side context config.
 class QuicServerTransportSocketFactory : public QuicTransportSocketFactoryBase {
 public:
-  QuicServerTransportSocketFactory(Ssl::ServerContextConfigPtr config)
-      : config_(std::move(config)) {}
+  QuicServerTransportSocketFactory(Stats::Scope& store, Ssl::ServerContextConfigPtr config)
+      : QuicTransportSocketFactoryBase(store, "server"), config_(std::move(config)) {}
+
+  void initialize() override {
+    config_->setSecretUpdateCallback([this]() {
+      // The callback also updates config_ with the new secret.
+      onSecretUpdated();
+    });
+  }
+
+  // Return TLS certificates if the context config is ready.
+  std::vector<std::reference_wrapper<const Envoy::Ssl::TlsCertificateConfig>>
+  getTlsCertificates() const {
+    if (!config_->isReady()) {
+      ENVOY_LOG(warn, "SDS hasn't finished updating Ssl context config yet.");
+      stats_.downstream_context_secrets_not_ready_.inc();
+      return {};
+    }
+    return config_->tlsCertificates();
```

Comment:
For my understanding when would this happen? Do we not block the usage of this on the SDS config being ready?

Reply:
I think we do block. But I saw similar logic in SSL socket to create a dummy socket so might as well added this.